clang/runtime/dpct-rt/include/dnnl_utils.hpp.inc

```diff
@@ -2147,6 +2295,49 @@ public:
 };
 // DPCT_LABEL_END
 
+// DPCT_LABEL_BEGIN|dropout_desc_def|dpct::dnnl
+// DPCT_DEPENDENCY_EMPTY
+// DPCT_CODE
+inline
+void dropout_desc::restore(engine_ext &engine, float p, void *state,
+                                  size_t state_size, unsigned long long seed) {
+  if (state) {
+    std::int64_t required_state_size = engine.get_dropout_state_size();
+    if (state_size < required_state_size) {
+      throw std::runtime_error("restore: state_size less than required state size.");
+    }
+    sycl::queue *q = engine.get_queue();
+    _imp->_p = p;
+    _imp->_seed = seed;
+    _imp->_state = state;
+    _imp->_host_state = std::vector<std::uint8_t>(required_state_size);
+    q->memcpy(_imp->_host_state.data(), _imp->_state, required_state_size).wait();
+    _imp->_rng_engine =
+        oneapi::mkl::rng::load_state<rng_engine_t>(
+            *q, _imp->_host_state.data());
+  }
+}
+
+inline
+void dropout_desc::set(engine_ext &engine, float p, void *state,
+                              size_t state_size, unsigned long long seed) {
+  _imp->_p = p;
+  if (state) {
+    std::int64_t required_state_size = engine.get_dropout_state_size();
+    if (state_size < required_state_size) {
+      throw std::runtime_error("set: no sufficient memory to save states.");
+    }
+    sycl::queue *q = engine.get_queue();
+    _imp->_seed = seed;
+    _imp->_state = state;
+    _imp->_host_state = std::vector<std::uint8_t>(required_state_size);
```

Comment:
could you clarify why the host_state is needed? From cuDNN API, it seems that all pointers to state are on device.
And from onemkl doc, it seems that if you create the rng_engine for a gpu device, the state pointer should be a device pointer as well (not sure though).

Reply:
Confirmed with oneMKL team, the state pointer should be a host pointer even the random generator engine is created for gpu device.